### PR TITLE
Creating a 'cloudproviders' package

### PR DESCRIPTION
**Pull request by hush-hush**
### What does this PR do?

The logic around cloudproviders is spread out all over the `util` package. This PR create a dedicated `cloudproviders` package.

### Motivation

Make it easier to add new cloudproviders.

### Describe how to test your changes

Install the agent on different cloudproviders and test that tags and hostname are still well reported.

Ensure that cluster name detection is working, too.

Also test that the metadata payload (printed in the logs in debug log level) is the same.

### Checklist

- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] The `need-change/operator` and `need-change/helm` labels has been applied if applicable.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated if applicable.

Note: Adding GitHub labels is only possible for contributors with write access.

### CODE REVIEW

File: pkg/util/cloudproviders/cloudproviders.go [hunk 1]
```diff
@@ -76,3 +82,59 @@ func GetCloudProviderNTPHosts(ctx context.Context) []string {
...
 }
+
+// GetHostAliases returns the hostname aliases from different provider
+// This should include:
+// * Alibaba
+// * Azure
+// * GCE
+// * cloudfoundry
+// * kubernetes
+// * Tencent
```
Comment (djmitche) [inline pkg/util/cloudproviders/cloudproviders.go:93]: I suspect this list will fall out of date!

File: pkg/otlp/model/attributes/azure/azure.go [hunk 2]
```diff
@@ -56,7 +56,7 @@ func HostnameFromAttributes(attrs pdata.AttributeMap) (string, bool) {
 // ClusterNameFromAttributes gets the Azure cluster name from attributes
 func ClusterNameFromAttributes(attrs pdata.AttributeMap) (string, bool) {
 	// Get cluster name from resource group
-	// https://github.com/DataDog/datadog-agent/blob/aad29b8/pkg/util/azure/azure.go#L51
+	// https://github.com/DataDog/datadog-agent/blob/aad29b8/pkg/util/cloudproviders/azure/azure.go#L51
```
Comment (djmitche) [inline pkg/otlp/model/attributes/azure/azure.go:59]: 404 :)

File: pkg/util/hostname.go [hunk 3]
```diff
@@ -179,7 +179,7 @@ func GetHostnameData(ctx context.Context) (HostnameData, error) {
 	configHostnameFilepath := config.Datadog.GetString("hostname_file")
 	if configHostnameFilepath != "" {
 		log.Debug("GetHostname trying `hostname_file` config option...")
-		if fileHostnameProvider, found := hostname.ProviderCatalog["file"]; found {
```
Comment (djmitche) [inline pkg/util/hostname.go:182]: Does it make sense to move this file to pkg/util/hostname?

File: pkg/util/hostname/providers.go [hunk 4]
```diff
@@ -5,15 +5,45 @@
...
+	if provider, found := providerCatalog[providerName]; found {
+		return provider
+	}
+	return nil
+}
+
+// GetHostname returns the hostname for a specific Provider if it was register
+func GetHostname(ctx context.Context, providerName string, options map[string]interface{}) (string, error) {
+	if provider, found := providerCatalog[providerName]; found {
+		log.Debugf("GetHostname trying %s trough API server...", providerName)
```
Comment (djmitche) [inline pkg/util/hostname/providers.go:38]:
```suggestion
		log.Debugf("GetHostname trying %s through API server...", providerName)
```

File: pkg/util/hostname/providers_test.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+func dummyProvider(ctx context.Context, options map[string]interface{}) (string, error) {
+	return "dummy-hostname", nil
+}
+
+func dummyErrorProvider(ctx context.Context, options map[string]interface{}) (string, error) {
+	return "", fmt.Errorf("Some error")
+}
+
+func dummyInvalideProvider(ctx context.Context, options map[string]interface{}) (string, error) {
```
Comment (djmitche) [inline pkg/util/hostname/providers_test.go:19]:
```suggestion
func dummyInvalidProvider(ctx context.Context, options map[string]interface{}) (string, error) {
```

**Review by djmitche — APPROVED** [review]

File: pkg/util/hostname.go (hunk 3 above)
Comment (hush-hush) [reply pkg/util/hostname.go:182]: I would very much so. I started looking into it and the change was massive. I'll open another PR dedicated to this.

File: pkg/util/hostname/providers.go [hunk 6]
```diff
@@ -5,15 +5,45 @@
...
+	if provider, found := providerCatalog[providerName]; found {
+		return provider
+	}
+	return nil
+}
+
+// GetHostname returns the hostname for a specific Provider if it was register
+func GetHostname(ctx context.Context, providerName string, options map[string]interface{}) (string, error) {
+	if provider, found := providerCatalog[providerName]; found {
+		log.Debugf("GetHostname trying %s through API server...", providerName)
```
Comment (sgnn7) [inline pkg/util/hostname/providers.go:38]: This message may need a bit tweaking as we won't always go through an API server (e.g. `file` provider)

File: pkg/util/hostname/providers.go [hunk 7]
```diff
@@ -5,15 +5,45 @@
...
+		name, err := provider(ctx, options)
+		if err != nil {
+			return "", err
+		}
+		if validate.ValidHostname(name) != nil {
+			return "", fmt.Errorf("Invalid hostname from %s provider", providerName)
```
Comment (sgnn7) [inline pkg/util/hostname/providers.go:44]: For debugging, it may be useful to show what that bad hostname was here.

**Review by sgnn7 — APPROVED** [review]
@hush-hush 2 tiny nits but LGTM. Great improvements all around!

File: pkg/otlp/model/attributes/azure/azure.go [hunk 8]
```diff
@@ -55,8 +55,7 @@ func HostnameFromAttributes(attrs pdata.AttributeMap) (string, bool) {
 
 // ClusterNameFromAttributes gets the Azure cluster name from attributes
 func ClusterNameFromAttributes(attrs pdata.AttributeMap) (string, bool) {
-	// Get cluster name from resource group
-	// https://github.com/DataDog/datadog-agent/blob/aad29b8/pkg/util/azure/azure.go#L51
+	// Get cluster name from resource group from pkg/util/cloudprovider/azure:GetClusterName
```
Comment (mx-psi) [inline pkg/otlp/model/attributes/azure/azure.go:58]: I would revert this change, so that the comment reflects the version we used when writing this code

File: pkg/util/hostname_docker.go [hunk 9]
```diff
@@ -14,41 +14,33 @@ import (
...
-	"github.com/DataDog/datadog-agent/pkg/util/hostname/validate"
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
 func getContainerHostname(ctx context.Context) string {
+	var err error
 	if config.IsFeaturePresent(config.Kubernetes) {
 		// Cluster-agent logic: Kube apiserver
-		if getKubeHostname, found := hostname.ProviderCatalog["kube_apiserver"]; found {
-			log.Debug("GetHostname trying Kubernetes trough API server...")
```
Comment (ahmed-mez) [inline pkg/util/hostname_docker.go:25]: could we keep this log line?

Comment (ahmed-mez) [reply pkg/util/hostname_docker.go:25]: oh I can see `GetHostname` is logging this instead, in this case we should get rid of `log.Debug("GetHostname trying Docker API...")` below to avoid duplicating the same log

File: pkg/otlp/model/attributes/azure/azure.go (hunk 8 above)
Comment (hush-hush) [reply pkg/otlp/model/attributes/azure/azure.go:58]: What the point of this comment ? I thought it was to reference the implementation. What the point of pointing to a piece of code that no longer exists ?

**Review by ahmed-mez — APPROVED** [review]
Mind adding a note in `Describe how to test your changes` about testing cluster name detection as well? thanks!

File: pkg/util/hostname.go (hunk 3 above)
Comment (hush-hush) [reply pkg/util/hostname.go:182]: It's here now; https://github.com/DataDog/datadog-agent/pull/9738/files#diff-b2d3d13fbbcae18918c7d8e4cb45f8b467b3a8b7e78294e0b05827ffda4fb6a6R38

File: pkg/otlp/model/attributes/azure/azure.go (hunk 8 above)
Comment (mx-psi) [reply pkg/otlp/model/attributes/azure/azure.go:58]: You are right, the implementation should not have breaking changes, approving from my team :)

**Review by mx-psi — APPROVED** [review]

**Review by robertjli — APPROVED** [review]

**Review by NouemanKHAL — APPROVED** [review]

**Comment by djmitche** [comment]
I'll rebase this shortly.  The conflict is with one of my PRs (and Maxime is away).

**Review by mfpierre — APPROVED** [review]

**Comment by djmitche** [comment]
It looks like the `integration_tests` failure is just an intermittent:
```
    cards_test.go:54: post: trace-agent not running
```
